File: Frameworks/RSCore/RSCore/NSPasteboard+RSCore.m
//
//  NSPasteboard+RSCore.m
//  RSCore
//
//  Created by Brent Simmons on 11/14/15.
//  Copyright © 2015 Ranchero Software, LLC. All rights reserved.
//

#import "NSPasteboard+RSCore.h"
#import "NSString+RSCore.h"


@implementation NSPasteboard (RSCore)

+ (nullable NSString *)rs_urlStringFromPasteboard:(NSPasteboard *)pasteboard {

	return [pasteboard rs_urlString];
}

- (nullable NSString *)rs_urlString {

	NSString *type = [self availableTypeFromArray:@[NSStringPboardType]];
	if (!type) {
		return nil;
	}

	NSString *s = [self stringForType:type];
	if (RSStringIsEmpty(s)) {
		return nil;
	}

	if ([s rs_stringMayBeURL]) {
		return s;
	}

	return nil;
}

@end

File: AppKitExtras/Sources/AppKitExtras/NSPasteboard+RSCore.swift
//
//  NSPasteboard+RSCore.swift
//  RSCore
//
//  Created by Brent Simmons on 2/11/18.
//  Copyright © 2018 Ranchero Software, LLC. All rights reserved.
//
#if os(macOS)
import AppKit

public extension NSPasteboard {

	@MainActor func copyObjects(_ objects: [Any]) {

		guard let writers = writersFor(objects) else {
			return
		}

		clearContents()
		writeObjects(writers)
	}

	func canCopyAtLeastOneObject(_ objects: [Any]) -> Bool {

		for object in objects {
			if object is PasteboardWriterOwner {
				return true
			}
		}
		return false
	}

}

public extension NSPasteboard {

	static func urlString(from pasteboard: NSPasteboard) -> String? {
		return pasteboard.urlString
	}

	private var urlString: String? {
		guard let type = self.availableType(from: [.string]) else {
			return nil
		}

		guard let str = self.string(forType: type), !str.isEmpty else {
			return nil
		}

		return str.mayBeURL ? str : nil
	}

}

private extension NSPasteboard {

	@MainActor func writersFor(_ objects: [Any]) -> [NSPasteboardWriting]? {

		let writers = objects.compactMap { ($0 as? PasteboardWriterOwner)?.pasteboardWriter }
		return writers.isEmpty ? nil : writers
	}
}
#endif
